File: examples/CBuilder/Simple/GUIControls/Renderers/fMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "fMain.h"
#include "RendererExampleClasses.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "BoldAbstractListHandle"
#pragma link "BoldAbstractModel"
#pragma link "BoldActions"
#pragma link "BoldControlPack"
#pragma link "BoldCursorHandle"
#pragma link "BoldDBActions"
#pragma link "BoldEdit"
#pragma link "BoldElements"
#pragma link "BoldGrid"
#pragma link "BoldHandle"
#pragma link "BoldHandleAction"
#pragma link "BoldHandles"
#pragma link "BoldListHandle"
#pragma link "BoldModel"
#pragma link "BoldNavigator"
#pragma link "BoldNavigatorDefs"
#pragma link "BoldPersistenceHandle"
#pragma link "BoldPersistenceHandleBDE"
#pragma link "BoldPersistenceHandleDB"
#pragma link "BoldRootedHandles"
#pragma link "BoldStringControlPack"
#pragma link "BoldSubscription"
#pragma link "BoldSystemHandle"
#pragma link "BoldAbstractDatabaseAdapter"
#pragma link "BoldAbstractPersistenceHandleDB"
#pragma link "BoldDatabaseAdapterIB"
#pragma link "BoldIBDatabaseAction"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCreate(TObject *Sender)
{
  Height = 330;
  Width = 447;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormCloseQuery(TObject *Sender, bool &CanClose)
{
  CanClose = true;
  if (BoldSystemHandle1->Active)
    if (BoldSystemHandle1->System->DirtyObjects->Count > 0)
      switch (MessageDlg("There are dirty objects. save them before exit?",
      mtConfirmation, TMsgDlgButtons() << mbYes << mbNo << mbCancel, 0))
      {
        case mrYes: BoldSystemHandle1->System->UpdateDatabase(); break;
        case mrNo: BoldSystemHandle1->System->Discard(); break;
        case mrCancel: CanClose = false;
      }
}
//---------------------------------------------------------------------------
void __fastcall TForm1::BoldActivateSystemAction1SystemOpened(
      TObject *Sender)
{
  if (blhGlobals->List->Count < 1)
    blhGlobals->List->AddNew();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::bsrSalaryLevelSetColor(TBoldElement *Element,
      TColor &AColor, int Representation, AnsiString Expression)
{
  if (Element)
    if (((TPerson*)Element)->Salary < ((TGlobals*)(blhGlobals->List->Elements[0]))->SalaryBreakPoint )
      AColor = clSilver;

 // if assigned(element) then
 //   if (element as TPerson).Salary < (blhGlobals.List[0] as TGlobals).SalaryBreakPoint then
 //    aColor := clSilver;

}
//---------------------------------------------------------------------------
void __fastcall TForm1::bsrSalaryLevelSubscribe(TBoldElement *Element,
      int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  TBoldExternalVariableList *var = NULL;
  Element->SubscribeToExpression("Globals.allInstances->first.salaryBreakPoint", Subscriber, false, false, var);
}
//---------------------------------------------------------------------------
AnsiString __fastcall TForm1::bsrFullNameGetAsString(TBoldElement *Element, int Representation, AnsiString Expression)
{
  AnsiString result = "";
  TPerson *person;

  if (Element)
  {
    person = dynamic_cast<TPerson*>(Element);
    if (person->FirstName != "")
      result = person->FirstName;
    else
      result = "<FirstName>";
    if (person->LastName != "")
      result += " " + person->LastName;
    else
      result += " <LastName>";
  }
  return result;
}
//---------------------------------------------------------------------------
bool __fastcall TForm1::bsrFullNameMayModify(TBoldElement *Element,
      int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  //int result = Element->EvaluateExpressionAsDirectElement("").ObserverMayModify(subscriber);
  bool result;
  TBoldExternalVariableList *var = NULL;
  TPerson *person;
  person = dynamic_cast<TPerson*>(Element);

  result = person->EvaluateExpressionAsDirectElement("firstName", var)->ObserverMayModify(Subscriber);
  result = (result && (person->EvaluateExpressionAsDirectElement("lastName", var)->ObserverMayModify(Subscriber)));

  return result;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::bsrFullNameSetAsString(TBoldElement *Element,
      AnsiString NewValue, int Representation, AnsiString Expression)
{
  TPerson *person = dynamic_cast<TPerson*>(Element);
  person->FirstName = NewValue.SubString(0, (NewValue.Pos(" ") - 1));
  person->LastName = NewValue.SubString(NewValue.Pos(" "), (NewValue.Length() - NewValue.Pos(" ") + 1));
}
//---------------------------------------------------------------------------
void __fastcall TForm1::bsrFullNameSubscribe(TBoldElement *Element,
      int Representation, AnsiString Expression,
      TBoldSubscriber *Subscriber)
{
  TPerson *person;
  TBoldExternalVariableList *var = NULL;

  person = dynamic_cast<TPerson*>(Element);
  person->SubscribeToExpression("firstName", Subscriber, false, false, var);
  person->SubscribeToExpression("lastName", Subscriber, false, false, var);
}
//---------------------------------------------------------------------------

File: examples/Delphi/Simple/GUIControls/Renderers/fMain.pas
unit fMain;

interface

uses
  SysUtils,
  Classes,
  Controls,
  Graphics,
  Forms,
  Dialogs,
  StdCtrls,
  ExtCtrls,
  ActnList,
  Grids,
  BoldElements,
  BoldHandle,
  BoldHandles,
  BoldSystemHandle,
  BoldRootedHandles,
  BoldAbstractListHandle,
  BoldCursorHandle,
  BoldListHandle,
  BoldPersistenceHandle,
  BoldPersistenceHandleDB,
  BoldModel,
  BoldControlPack,
  BoldStringControlPack,
  BoldSubscription,
  BoldGrid,
  BoldNavigator,
  BoldEdit,
  BoldHandleAction,
  BoldActions,
  BoldDBActions,
  BoldAbstractModel,
  BoldNavigatorDefs, DB, IBDatabase, BoldAbstractDatabaseAdapter,
  BoldDatabaseAdapterIB, BoldAbstractPersistenceHandleDB,
  BoldIBDatabaseAction, ComCtrls, BoldTrackBar, BoldExpressionHandle,
  BoldLabel, BoldReferenceHandle;

type
  TForm1 = class(TForm)
    BoldModel1: TBoldModel;
    BoldSystemHandle1: TBoldSystemHandle;
    BoldSystemTypeInfoHandle1: TBoldSystemTypeInfoHandle;
    blhPerson: TBoldListHandle;
    BoldGrid1: TBoldGrid;
    bsrSalaryLevel: TBoldAsStringRenderer;
    BoldNavigator1: TBoldNavigator;
    bsrFullName: TBoldAsStringRenderer;
    lblSalaryBreakPoint: TLabel;
    ActionList1: TActionList;
    BoldActivateSystemAction1: TBoldActivateSystemAction;
    BoldIBDatabaseAction1: TBoldIBDatabaseAction;
    Button1: TButton;
    Button2: TButton;
    BoldPersistenceHandleDB1: TBoldPersistenceHandleDB;
    BoldDatabaseAdapterIB1: TBoldDatabaseAdapterIB;
    IBDatabase1: TIBDatabase;
    BoldTrackBar1: TBoldTrackBar;
    Label1: TLabel;
    BoldTrackBar2: TBoldTrackBar;
    BoldLabel1: TBoldLabel;
    brhGlobals: TBoldReferenceHandle;
    procedure bsrSalaryLevelSetColor(Element: TBoldElement;
      var AColor: TColor; Representation: Integer; Expression: String);
    procedure bsrSalaryLevelSubscribe(Element: TBoldElement;
      Representation: Integer; Expression: String;
      Subscriber: TBoldSubscriber);
    function bsrFullNameGetAsString(Element: TBoldElement;
      Representation: Integer; Expression: String): String;
    procedure bsrFullNameSetAsString(Element: TBoldElement;
      NewValue: String; Representation: Integer; Expression: String);
    procedure bsrFullNameSubscribe(Element: TBoldElement;
      Representation: Integer; Expression: String;
      Subscriber: TBoldSubscriber);
    function bsrFullNameMayModify(Element: TBoldElement;
      Representation: Integer; Expression: String;
      Subscriber: TBoldSubscriber): Boolean;
    procedure FormCreate(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure BoldActivateSystemAction1SystemOpened(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Form1: TForm1;

implementation

uses
  RendererExampleClasses;

{$R *.DFM}

// If Salary is less than BreakPoint, set color to silver.
procedure TForm1.bsrSalaryLevelSetColor(Element: TBoldElement;
  var AColor: TColor; Representation: Integer; Expression: String);
begin
  if assigned(element) then
    if (element as TPerson).Salary < (brhGlobals.value as TGlobals).SalaryBreakPoint then
      aColor := clSilver;
end;

procedure TForm1.bsrSalaryLevelSubscribe(Element: TBoldElement;
  Representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber);
begin
  // subscribe to changes in the salary
  Element.SubscribeToExpression('salary', Subscriber, False);
  // Subscribe to changes in BreakPoint value.
  BoldSystemHandle1.System.SubscribeToExpression('Globals.allInstances->first.salaryBreakPoint', Subscriber, False);
end;

// Assemble a FullName from First and Last.
function TForm1.bsrFullNameGetAsString(Element: TBoldElement;
  Representation: Integer; Expression: String): String;
begin
  Result := '';
  if Assigned(Element) then
  begin
    with Element as TPerson do
    begin
      if FirstName <> '' then
        Result := FirstName
      else
        Result := '<FirstName>';
      if LastName <> '' then
        Result := Result + ' ' + LastName
      else
        Result := Result + ' <LastName>';
    end;
  end;
end;

// Break the FullName at the space and set Firstname and LastName.
procedure TForm1.bsrFullNameSetAsString(Element: TBoldElement;
  NewValue: String; Representation: Integer; Expression: String);
begin
  with element as TPerson do
  begin
    FirstName := Copy(NewValue, 0, (Pos(' ', NewValue) -1));
    LastName  := Copy(NewValue, Pos(' ', NewValue), (Length(NewValue) -
                     Pos(' ', NewValue) +1));
  end;
end;

// Subscribe to changes in FirstName and LastName values.
procedure TForm1.bsrFullNameSubscribe(Element: TBoldElement;
  Representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber);
begin
  with Element as TPerson do
  begin
    SubscribeToExpression('firstName', Subscriber, False);
    SubscribeToExpression('lastName', Subscriber, False);
  end;
end;

// Check if its OK to change FirstName and LastName.
function TForm1.bsrFullNameMayModify(Element: TBoldElement;
  Representation: Integer; Expression: String;
  Subscriber: TBoldSubscriber): Boolean;
begin
  with Element as TPerson do
  begin
    result := EvaluateExpressionAsDirectElement('firstName').ObserverMayModify( subscriber );
    result := result and EvaluateExpressionAsDirectElement('lastName').ObserverMayModify( subscriber );
  end;
end;

// Set size of form before show.
procedure TForm1.FormCreate(Sender: TObject);
begin
  AutoSize := true;
end;

// Give the user the opportunity to save changes before close.
procedure TForm1.FormCloseQuery(Sender: TObject; var CanClose: Boolean);
begin
  CanClose := True;
  if BoldSystemHandle1.Active then
    if BoldSystemHandle1.System.DirtyObjects.Count > 0 then
      case MessageDlg( 'There are dirty objects. save them before exit?', mtConfirmation, [mbYes, mbNo, mbCancel], 0) of
        mrYes: BoldSystemHandle1.System.UpdateDatabase;
        mrNo: BoldSystemHandle1.System.Discard;
        mrCancel: CanClose := False;
      end;
end;

procedure TForm1.BoldActivateSystemAction1SystemOpened(Sender: TObject);
var
  Globals: TGlobals;
begin
  if BoldSystemhandle1.System.ClassByExpressionName['Globals'].Count > 0 then
    Globals :=BoldSystemhandle1.System.ClassByExpressionName['Globals'][0] as TGlobals
  else
    Globals := TGlobals.Create(BoldSystemHandle1.System);
  brhGlobals.Value := Globals;
end;

end.


